### PR TITLE
Delete graphite data when resource gets deleted

### PR DESCRIPTION
**Pull request by rishubhjain**
Tendrl-bug-id: Tendrl/monitoring-integration#148

### CODE REVIEW

**Comment by rishubhjain** [comment]
@r0h4n @shtripat Please review, It is tested

File: tendrl/monitoring_integration/flows/update_graphite/__init__.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+from tendrl.commons import flows
+from tendrl.commons.utils import log_utils as logger
+from tendrl.commons.utils import etcd_utils
+from tendrl.monitoring_integration.grafana import create_dashboards
+
+
+WHISPER_PATH = "/var/lib/carbon/whisper/tendrl/"
+
+class UpdateGraphite(flows.BaseFlow):
```
Comment (r0h4n) [inline tendrl/monitoring_integration/flows/update_graphite/__init__.py]: Call this "DeleteResourceFromGraphite"

File: tendrl/monitoring_integration/flows/update_graphite/__init__.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import datetime
+
+
+from tendrl.commons import flows
+from tendrl.commons.utils import log_utils as logger
+from tendrl.commons.utils import etcd_utils
+from tendrl.monitoring_integration.grafana import create_dashboards
+
+
+WHISPER_PATH = "/var/lib/carbon/whisper/tendrl/"
```
Comment (r0h4n) [inline tendrl/monitoring_integration/flows/update_graphite/__init__.py]: Is this path the same for all versions of carbon/graphite? Can you read this path from the Graphite/Carbon conf file?

File: tendrl/monitoring_integration/objects/definition/monitoring_integration.yaml [hunk 3]
```diff
@@ -29,6 +29,20 @@ namespace.monitoring:
       type: Create
       version: 1
       uuid: 1951e821-7aa9-4a91-8183-e73bc8275bde
+    UpdateGraphite:
```
Comment (r0h4n) [inline tendrl/monitoring_integration/objects/definition/monitoring_integration.yaml]: DeleteResourceFromGraphite

File: tendrl/monitoring_integration/flows/update_graphite/__init__.py (hunk 2 above)
Comment (rishubhjain) [reply tendrl/monitoring_integration/flows/update_graphite/__init__.py]: I will retrieve the path from https://github.com/Tendrl/monitoring-integration/blob/master/etc/tendrl/monitoring-integration/graphite/carbon.conf.sample#L34, after it gets copied to the desired location

**Comment by rishubhjain** [comment]
@shtripat I have made the changes, Please review

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 4]
```diff
@@ -20,11 +19,30 @@ def run(self):
             self.parameters.get("Trigger.resource_name")).lower()
         resource_type = str(
             self.parameters.get("Trigger.resource_type")).lower()
-        import pdb; pdb.set_trace()
         self.update_graphite(cluster_id, resource_name,
                              resource_type.lower())
 
+    def get_path(self):
+        carbon_path = "/etc/tendrl/monitoring-integration/carbon.conf"
```
Comment (shtripat) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: Is this file a ini format? If so cant we load ini to dict and refer fields.

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 5]
```diff
@@ -20,11 +19,30 @@ def run(self):
             self.parameters.get("Trigger.resource_name")).lower()
         resource_type = str(
             self.parameters.get("Trigger.resource_type")).lower()
-        import pdb; pdb.set_trace()
         self.update_graphite(cluster_id, resource_name,
                              resource_type.lower())
 
+    def get_path(self):
```
Comment (shtripat) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: call this `get_data_dir_path` may be

**Comment by rishubhjain** [comment]
@shtripat Made the necessary changes, Please review

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            return
+        if resource_type == "volume":
+            resource_path = os.path.join(WHISPER_PATH, "clusters",
+                                         str(cluster_id),
+                                         "volumes", resource_name)
+            archive_path = os.path.join(WHISPER_PATH, "clusters",
+                                        str(cluster_id),
+                                        "archive", "volumes")
+            if not os.path.exists(archive_path):
+                os.system("mkdir -p " + str(archive_path))
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: os module has "makedirs" function, you can directly use that.
example: os.makedirs("a/b/c")

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        carbon_config.read(carbon_path)
+        try:
+            whisper_path = str(carbon_config.get("cache", "local_data_dir")
+            whisper_path = os.path.join(whisper_path, "tendrl/")
+            return whisper_path
+        except KeyError:
+            return None
+
+
+    def update_graphite(self, cluster_id, resource_name, resource_type):
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: can we divide this function into submodules? The function is way too long, which makes it difficult to read, understand and review.

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                    logger.log("info", NS.get("publisher_id", None),
+                              {'message': "Brick " + str(resource_name) + 
+                                       "deleted from graphite"})
+                else:
+                     logger.log("error", NS.get("publisher_id", None),
+                               {'message': "Brick " + str(resource_name) + 
+                                       "deletion from graphite failed"})
+            archive_path = os.path.join(archive_path,
+                                        "volumes", vol_name)
+            os.system("mkdir -p " + str(archive_path))
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: same as above

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                                volume_affected_list.append(volume_name)
+                            except etcd.EtcdKeyNotFound:
+                                pass
+                            flag = True
+                            continue
+            archive_path = os.path.join(WHISPER_PATH, "clusters",
+                                        str(cluster_id),
+                                        "archive", "nodes")
+            if not os.path.exists(archive_path):
+                os.system("mkdir -p " + str(archive_path))
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: same as above

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                               {'message': "Host " + str(resource_name) + 
+                                       "deleted from graphite"})
+                else:
+                     logger.log("error", NS.get("publisher_id", None),
+                                {'message': "Host " + str(resource_name) + 
+                                       "deletion from graphite failed"})
+            for vol_name in volume_affected_list:
+                archive_path = os.path.join(archive_path,
+                                            "volumes", vol_name)
+                os.system("mkdir -p " + str(archive_path))
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: same as above

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                            continue
+            archive_path = os.path.join(WHISPER_PATH, "clusters",
+                                        str(cluster_id),
+                                        "archive", "nodes")
+            if not os.path.exists(archive_path):
+                os.system("mkdir -p " + str(archive_path))
+            resource_folder_name = str(host_name) + "_" + \
+                str(datetime.datetime.now().isoformat())
+            archive_path = os.path.join(archive_path, resource_folder_name)
+            os.system("mkdir -p " + str(archive_path))
```
Comment (cloudbehl) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: same as above

**Comment by rishubhjain** [comment]
@cloudbehl I have made the changes, Please review

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 12]
```diff
@@ -30,179 +30,186 @@ def get_data_dir_path(self):
...
+            flag = False
+            single_volume_key = os.path.join(volume_key, volume_id,
+                                             "Bricks")
+            subvolume_list = create_dashboards.get_resource_keys("", single_volume_key)
+            for subvolume in subvolume_list:
+                subvolume_brick_key = os.path.join(single_volume_key,
+                                                   subvolume)
+                brick_list = create_dashboards.get_resource_keys("", subvolume_brick_key)
+                if flag:
+                    flag = False
```
Comment (shtripat) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: Where is flag being set as `True`?

Comment (rishubhjain) [reply tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: @shtripat https://github.com/Tendrl/monitoring-integration/pull/154/files#diff-df788ce9a7b7a09945c3cd9e735da7d5R159

Comment (shtripat) [reply tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: Ah my bad didnt check down enough :) Ack

File: tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+from tendrl.commons.utils import etcd_utils
+from tendrl.monitoring_integration.grafana import create_dashboards
+
+
+
+class DeleteResourceFromGraphite(flows.BaseFlow):
+
+    def run(self):
+        super(UpdateGraphite, self).run()
+        cluster_id = self.parameters.get("TendrlContext.integration_id")
```
Comment (r0h4n) [inline tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: this is not cluster_id, stop adding confusing code and use integration_id everywhere

Comment (rishubhjain) [reply tendrl/monitoring_integration/flows/delete_resource_from_graphite/__init__.py]: @r0h4n Made the required changes, Will make changes in other files also in separate PR

**Pull request merged**